### PR TITLE
redo-log: add basic framework for redo log manager and writer

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 

This PR adds basic logic for #2351 to complete subtasks 4-2, 5-1 and 5-2, and implements the following workflows.

- Add interface of redo log writer and basic redo log manager.
- Add necessary changefeed config, currently it is simple, containing consistent level and consistent storage.
- Implement basic watermark mechanism, adding `RedoLogTs` in each table sink.
- Add a mock blackhole redo log writer and verify the replication can run normally with consistent replication feature enabled.

There still exists some remaining works, including

   - [ ] Add admin command, such as remove a table from redo log writer
   - [ ] Better interface abstraction
   - [x] Unit test for new code

### What is changed and how it works?

   As described in PR description

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/redo/storage.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	Index [][]int                `json:"index"`
+}
+
+// LogWriter is a writer abstraction for redo log storage layer
+type LogWriter interface {
+	WriteLog(ctx context.Context, tableID model.TableID, log []*RowRedoLog) (offset uint64, err error)
+	// FlushLog sends resolved-ts from table pipeline to log writer, it is
+	// essential to flush when a table doesn't have any row change event for
+	// sometime, and the resolved ts of this table should be moved forward.
+	FlushLog(ctx context.Context, tableID model.TableID, ts uint64) error
```
Comment (overvenus) [inline cdc/redo/storage.go:34]:
```suggestion
	FlushLog(ctx context.Context, tableID model.TableID, resolvedTs uint64) error
```

File: cdc/redo/storage.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"context"
+
+	"github.com/pingcap/ticdc/cdc/model"
+)
+
+// RowRedoLog stores one RowChangedEvent and its index column
+type RowRedoLog struct {
+	Row   *model.RowChangedEvent `json:"row"`
+	Index [][]int                `json:"index"`
```
Comment (overvenus) [inline cdc/redo/storage.go:25]: Could you add some comments about why `Index [][]int` is needed?

File: cdc/redo/manager.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+const (
+	consistentLevelNomal    consistentLevelType = "normal"
+	consistentLevelEventual consistentLevelType = "eventual"
+)
+
+type consistentStorage string
+
+const (
+	consistentStorageLocal     consistentStorage = "local"
+	consistentStorageS3        consistentStorage = "s3"
```
Comment (overvenus) [inline cdc/redo/manager.go:44]:
```suggestion
	// consistentStorageS3        consistentStorage = "s3"
```

Please do not add it until TiCDC actually supports it.

File: cdc/redo/manager.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+)
+
+var updateOffsetInterval = time.Second
+
+type consistentLevelType string
+
+const (
+	consistentLevelNomal    consistentLevelType = "normal"
+	consistentLevelEventual consistentLevelType = "eventual"
+)
```
Comment (overvenus) [inline cdc/redo/manager.go:38]: Why do we need level? What's the difference between normal and eventual?

File: cdc/redo/manager.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+		return m.tableIDs[i] >= tableID
+	})
+	if i < len(m.tableIDs) && m.tableIDs[i] == tableID {
+		copy(m.tableIDs[i:], m.tableIDs[i+1:])
+		m.tableIDs = m.tableIDs[:len(m.tableIDs)-1]
+		delete(m.offsets, tableID)
+	} else {
+		log.Warn("remove a table not maintained in redo log manager", zap.Int64("table-id", tableID))
+	}
+	// TODO: send remove table command to redo log writer
```
Comment (overvenus) [inline cdc/redo/manager.go:186]: Do we need a new method in `LogWriter`?

File: cdc/redo/storage.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// LogWriter is a writer abstraction for redo log storage layer
+type LogWriter interface {
+	WriteLog(ctx context.Context, tableID model.TableID, log []*RowRedoLog) (offset uint64, err error)
```
Comment (overvenus) [inline cdc/redo/storage.go:30]: What is the offset?

File: cdc/redo/manager.go (hunk 5 above)
Comment (amyangfei) [reply cdc/redo/manager.go:186]: Add it when we really need

File: cdc/redo/manager.go (hunk 4 above)
Comment (amyangfei) [reply cdc/redo/manager.go:38]: normal is default mode, without eventually consistent guarantee

File: cdc/redo/manager.go (hunk 3 above)
Comment (amyangfei) [reply cdc/redo/manager.go:44]: It is a prototype PR, the first redo log writer we will add is s3, IMHO it is ok to leave it here

File: cdc/redo/storage.go (hunk 6 above)
Comment (amyangfei) [reply cdc/redo/storage.go:30]: will be changed to resolved ts

**Pull request closed without merging**